lib/tree.rb

```diff
@@ -1,64 +1,149 @@
+    return TreeNode.new(key, value) if current_node.nil?
+    
+    if key < current_node.key
+      if current_node.left.nil?
+        current_node.left = add_helper(current_node.left, key, value)
+      else
+        add_helper(current_node.left, key, value)
+      end
+    else
+      if current_node.right.nil?
+        current_node.right = add_helper(current_node.right, key, value)
+      else
+        add_helper(current_node.right, key, value)
+      end
+    end
+    
+  end
+  
   def add(key, value)
-    raise NotImplementedError
+    if @root.nil?
+      @root = add_helper(@root, key, value)
+    else
+      add_helper(@root, key, value)
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+    key > current_node.key ? find_helper(current_node.right, key) : find_helper(current_node.left, key)
+  end
+  
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+  end
+  
   def inorder
-    raise NotImplementedError
+    inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+  end
+  
   def preorder
-    raise NotImplementedError
+    preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+    
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << {key: current_node.key, value: current_node.value}
+  end
+  
   def postorder
-    raise NotImplementedError
+    postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def height_helper(current_node, max_levels_count, levels_count)
+    return max_levels_count if current_node.nil?
+    
+    levels_count += 1
+    if levels_count > max_levels_count
+      max_levels_count = levels_count
+    end
+    
+    height_helper(current_node.left, max_levels_count, levels_count)
+    height_helper(current_node.right, max_levels_count, levels_count)
```

Comment:
This isn't working because you're **always** returning the right subtree's height.